src/frontend/actors_generation_phase.nit

```diff
@@ -189,12 +186,46 @@ redef class Proxy{{{classname}}}
 end
 """)
 
-		redef_classes.add(
+	if mclassdef.is_intro then redef_classes.add(
 """
 redef class {{{classname}}}
-redef var async is lazy do return new Proxy{{{classname}}}.proxy(self)
+	var m = new Mutex
+	var lazy_proxy: Proxy{{{classname}}} is lazy do return new Proxy{{{classname}}}.proxy(self)
+
+	redef fun async: Proxy{{{classname}}} do
+		m.lock
+		var p = lazy_proxy
+		m.unlock
+		return p
+	end
 end
 """)
+
+	end
+
+	redef fun process_nmodule(nmodule) do
+		var mmod = nmodule.mmodule
+		if mmod == null then return
+
+		if mmod.name.has("actors_") then return
```

Comment:
Is this to avoid the generated module?

If so you would skip all other modules beginning with `actors_`, which could be common in client code using actors (this very module starts with `actors_`). A trick would be to add an illegal character in the name (like space), so it can only be generated and not used by a programmer. Or check if `generated_actor_modules.has(name)`, or set a boolean in the `AModule`...